cache_flat_mutation_reader.hh

```diff
@@ -730,20 +749,75 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
-        if (_next_row.range_tombstone_for_row() != _current_tombstone) [[unlikely]] {
-            auto tomb = _next_row.range_tombstone_for_row();
-            auto new_lower_bound = position_in_partition::before_key(_next_row.position());
-            clogger.trace("csm {}: rtc({}, {})", fmt::ptr(this), new_lower_bound, tomb);
-            push_mutation_fragment(mutation_fragment_v2(*_schema, _permit, range_tombstone_change(new_lower_bound, tomb)));
-            _lower_bound = std::move(new_lower_bound);
-            _current_tombstone = tomb;
-            _read_context.cache()._tracker.on_range_tombstone_read();
+        bool remove_row = false;
+
+        if (!_next_row.dummy()
+            && _next_row.at_a_row()
+            && _next_row.continuous()
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+
+            deletable_row& row = _next_row.latest_row();
+            tombstone range_tomb = _next_row.range_tombstone_for_row();
+            auto t = row.deleted_at();
+            t.apply(range_tomb);
+
+            auto tomb_expired = [&](row_tombstone tomb) {
+                return (t && t.max_deletion_time() < _gc_before);
+            };
+
+            auto is_row_dead = [&](const deletable_row& row) {
+                return (!row.marker().is_missing() && row.marker().is_dead(_read_time));
+            };
+
+            if (tomb_expired(t) || is_row_dead(row)) {
+                can_gc_fn always_gc = [](tombstone) { return true; };
+                // compact_and_expire can change marker state so we have to keep its value and restore back if row isn't gonna be removed
+                row_marker marker = row.marker();
+
+                with_allocator(_snp->region().allocator(), [&] {
+                    if (!row.compact_and_expire(*_schema, range_tomb, _read_time, always_gc, _gc_before, nullptr)) {
+                        remove_row = row.empty();
+                    }
+                });
+
+                if (!remove_row) {
+                    row.marker() =  marker;
+                }
```

Comment:
In this case, each read will pay the cost of trying to compact the row, which will be futile. This is an argument against restoring the marker.

Reply:
I agree. I'd rather use better solution if had one. As I told at least one test expects not compacted marker:
`[ FAIL ] boost.row_cache_test.test_row_cache_conforms_to_mutation_source`

Reply:
Let's disable compaction of markers by default like we did for tombstones, can be based on the same same thing (gc_state).

Reply:
Maybe we could just call row.compact_and_expire() only when gc_state is set?

Reply:
Maybe we should just don't perform cache compaction when gc_state is not set?
And restore markers on bad_alloc exception only?

Reply:
Yes (except that we should catch all exceptions).

There is row_cache_test.cc::test_concurrent_reads_and_eviction which is a stress test, it would be good to run it with gc_state to exercise compaction path.